src/shared/fork-journal.c

```diff
@@ -0,0 +1,217 @@
+
+        /* And let's then fail the whole thing, because regardless what the exit status of the child is
+         * (i.e. even if successful), if it exits before sending READY=1 something is wrong. */
+
+        return log_debug_errno(SYNTHETIC_ERRNO(EPROTO), "Child " PID_FMT " died before sending notification message.", child->pid);
+}
+
+static int on_child_notify(sd_event_source *s, int fd, uint32_t revents, void *userdata) {
+        PidRef *child = ASSERT_PTR(userdata);
+        int r;
+
+        _cleanup_strv_free_ char **msg = NULL;
+        _cleanup_(pidref_done) PidRef sender = PIDREF_NULL;
+        r = notify_recv_strv(fd, &msg, /* ret_ucred= */ NULL, &sender);
+        if (r == -EAGAIN)
+                return 0;
+        if (r < 0)
+                return r;
+
+        if (!pidref_equal(child, &sender)) {
+                log_warning("Received notification message from unexpected process " PID_FMT " (expected " PID_FMT "), ignoring.",
+                            sender.pid, child->pid);
+                return 0;
+        }
+
+        if (strv_find(msg, "READY=1"))
+                return sd_event_exit(sd_event_source_get_event(s), EXIT_SUCCESS);
+
+        const char *e = strv_find_startswith(msg, "ERRNO=");
+        if (e) {
+                int error;
+
+                r = safe_atoi(e, &error);
+                if (r < 0) {
+                        log_debug_errno(r, "Received invalid ERRNO= notification message, ignoring: %m");
+                        return 0;
+                }
+                if (error <= 0) {
+                        log_debug("Received non-positive ERRNO= notification message, ignoring: %m");
+                        return 0;
+                }
+
+                return -error;
+        }
+
+        return 0;
+}
+
+int journal_fork(RuntimeScope scope, char **units, PidRef *ret_pidref) {
+        int r;
+
+        assert(scope >= 0);
+        assert(scope < _RUNTIME_SCOPE_MAX);
+        assert(ret_pidref);
+
+        if (!is_main_thread())
+                return -EPERM;
+
+        if (strv_isempty(units))
+                return 0;
+
+        _cleanup_close_ int notify_fd = socket(AF_UNIX, SOCK_DGRAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (notify_fd < 0)
+                return log_debug_errno(errno, "Failed to allocate AF_UNIX socket for notifications: %m");
+
+        r = setsockopt_int(notify_fd, SOL_SOCKET, SO_PASSCRED, true);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to enable SO_PASSCRED: %m");
+
+        r = setsockopt_int(notify_fd, SOL_SOCKET, SO_PASSPIDFD, true);
+        if (r < 0)
+                log_debug_errno(r, "Failed to enable SO_PASSPIDFD, ignoring: %m");
+
+        /* Pick an address via auto-bind */
+        _cleanup_free_ char *addr_string = NULL;
+        r = socket_autobind(notify_fd, &addr_string);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind AF_UNIX socket: %m");
```

Comment:
notify_socket_prepare()

Maybe teach it to optionally return the event source instead of making it "floating"? Or, call `sd_event_exit` everywhere manually?